src/Npgsql/PreparedStatement.cs

```diff
@@ -112,7 +112,8 @@ internal void CompletePrepare()
         internal void CompleteUnprepare()
         {
             _manager.BySql.Remove(Sql);
-            _manager.NumPrepared--;
+            if (IsPrepared || State == PreparedState.BeingUnprepared)
```

Comment:
State is only BeingUnprepared when NpgsqlCommand.Unprepare is called - that method is currently lacking proper error handling in general (opened #3125 to track). Unless I'm mistaken, in that case, failure should actually *not* decrease NumPrepared (because unprepare has failed), so this check is inaccurate.

So unless I'm missing something, I'd rather we removed the BeingUnprepared check and handle the whole scenario separately at some point as part of #3125 (note that Unprepare is an extremely rare-used method...).

Reply:
Am I understanding you right, the correct fix is to leave only IsPrepared check and not decrease NumPrepared for Unprepare()?
This will cause DeriveParametersForAutoPreparedStatement test to fail. So I tweak the test and the rest will be fixed in #3125?

Reply:
OK, I see the test failure. I'll go ahead and merge this for now, and take care of the rest in #3125, thanks!